server/util/rpcutil/rpcutil.go

```diff
@@ -0,0 +1,123 @@
+package rpcutil
+
+import (
+	"context"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+)
+
+type StreamMsg[T proto.Message] struct {
+	Data  T
+	Error error
+}
+
+type RecvStream[T proto.Message] interface {
+	Recv() (T, error)
+}
+
+type Receiver[T proto.Message] struct {
+	ctx      context.Context
+	recvChan chan StreamMsg[T]
+}
+
+// RecvWithTimeoutCause waits for a message on the underlying stream, waiting
+// a maximum of timeout. If timeout is reached, the given cause is returned
+// as the error.
+func (r *Receiver[T]) RecvWithTimeoutCause(timeout time.Duration, cause error) (T, error) {
+	ctx, cancel := context.WithTimeoutCause(r.ctx, timeout, cause)
+	defer cancel()
+	select {
+	case msg := <-r.recvChan:
+		return msg.Data, msg.Error
+	case <-ctx.Done():
+		return *new(T), ctx.Err()
+	}
+}
+
+// NewReceiver returns a stream handle that can be used to implement more
+// advanced stream handling, such as per-receive timeouts.
+//
+// Example usage:
+//
+// receiver := rpcutil.NewReceiver[*bspb.ReadResponse](ctx, stream)
+//
+//	for {
+//		rsp, err := receiver.RecvWithTimeoutCause(5 * time.Second, status.DeadlineExceededError("blah blah blah"))
+//		// handle rsp and err
+//	}
+func NewReceiver[T proto.Message](ctx context.Context, stream RecvStream[T]) Receiver[T] {
+	streamMsgs := make(chan StreamMsg[T])
+	go func() {
+		for {
+			rsp, err := stream.Recv()
+			select {
+			case streamMsgs <- StreamMsg[T]{rsp, err}:
+			case <-ctx.Done():
+				return
+			}
+			if err != nil {
+				return
+			}
+		}
+	}()
+	return Receiver[T]{ctx, streamMsgs}
+}
+
+type SendStream[T proto.Message] interface {
+	Send(T) error
+}
+
+type Sender[T proto.Message] struct {
+	ctx      context.Context
+	sendChan chan T
+	errChan  chan error
+}
+
+// SendWithTimeoutCause attempts to send a message on the underlying stream,
+// waiting a maximum of timeout. If timeout is reached, the given cause is
+// returned as the error.
+//
+// Note that gRPC sends are asynchronous in the sense that the protocol does not
+// acknowledge individual messages. A timeout wil only occur if the sender
+// exhausts the flow-control window and the receiver does not increase it.
+func (s *Sender[T]) SendWithTimeoutCause(msg T, timeout time.Duration, cause error) error {
+	s.sendChan <- msg
```

Comment:
I think this needs a `select {}` with `<-s.ctx.Done()` here, otherwise it could get stuck?

Reply:
sendChan is a buffered channel and SendWithTimeoutCause does not return until either we have received an err and the channel is empty or the context is done and SendWithTimeoutCause won't be called again.